cpp/src/explainer/tree_shap.cu

```diff
@@ -0,0 +1,326 @@
+              if (b.path_idx < a.path_idx) return false;
+
+              if (a.feature_idx < b.feature_idx) return true;
+              if (b.feature_idx < a.feature_idx) return false;
+              return false;
+            });
+
+  for (auto i = 0ull; i < tmp.size(); i++) {
+    auto e = tmp[i];
+    if (i == 0 || e.path_idx != tmp[i - 1].path_idx) {
+      os << "path_idx:" << e.path_idx << ", leaf value:" << e.v;
+      os << "\n";
+    }
+    os << " (feature:" << e.feature_idx << ", pz:" << e.zero_fraction << ", ["
+       << e.split_condition.feature_lower_bound << "<=x<"
+       << e.split_condition.feature_upper_bound << "])";
+    os << "\n";
+  }
+  return os;
+}
+
+class DenseDatasetWrapper {
+  const float* data;
+  std::size_t num_rows;
+  std::size_t num_cols;
+
+ public:
+  DenseDatasetWrapper() = default;
+  DenseDatasetWrapper(const float* data, int num_rows, int num_cols)
+      : data(data), num_rows(num_rows), num_cols(num_cols) {}
+  __device__ float GetElement(std::size_t row_idx, std::size_t col_idx) const {
+    return data[row_idx * num_cols + col_idx];
+  }
+  __host__ __device__ std::size_t NumRows() const { return num_rows; }
+  __host__ __device__ std::size_t NumCols() const { return num_cols; }
+};
+
+
+}  // anonymous namespace
+
+namespace ML {
+namespace Explainer {
+
+template <typename ThresholdType, typename LeafType>
+ExtractedPathHandle
+extract_paths_impl(const tl::ModelImpl<ThresholdType, LeafType>& model) {
+  if (!std::is_same<ThresholdType, LeafType>::value) {
+    RAFT_FAIL("ThresholdType and LeafType must be identical");
+  }
+  if (model.task_type != tl::TaskType::kBinaryClfRegr
+      && model.task_type != tl::TaskType::kMultiClfGrovePerClass) {
+    RAFT_FAIL("Only tree models with task type kBinaryClfRegr, "
+              "kMultiClfGrovePerClass are supported. So for example, "
+              "scikit-learn trees are not supported.");
+  }
+  std::unique_ptr<ExtractedPath> path_container
+    = ExtractedPath::Create<ThresholdType>();
+  ExtractedPathImpl<ThresholdType>* paths
+    = dynamic_cast<ExtractedPathImpl<ThresholdType>*>(path_container.get());
+
+  std::size_t path_idx = 0;
+  int tree_idx = 0;
+  int num_groups = 1;
+  if (model.task_type == tl::TaskType::kMultiClfGrovePerClass
+      && model.task_param.num_class > 1) {
+    num_groups = model.task_param.num_class;
+  }
+  for (const tl::Tree<ThresholdType, LeafType>& tree : model.trees) {
+    std::vector<int> parent_id(tree.num_nodes, -1);
+    // Compute parent ID of each node
+    for (int i = 0; i < tree.num_nodes; i++) {
+      if (!tree.IsLeaf(i)) {
+        parent_id[tree.LeftChild(i)] = i;
+        parent_id[tree.RightChild(i)] = i;
+      }
+    }
+
+    // Find leaf nodes
+    // Work backwards from leaf to root, order does not matter
+    // It's also possible to work from root to leaf
+    for (int i = 0; i < tree.num_nodes; i++) {
+      if (tree.IsLeaf(i)) {
+        float v = static_cast<float>(tree.LeafValue(i));
```

Comment:
The type is already unambiguously shown in the line; I'd recommend deduplicating it to keep a single source of truth for the type.
```suggestion
        auto v = static_cast<float>(tree.LeafValue(i));
```